### ISSUE TITLE
TFLite -- Converting the trained pb file into tflite failed predictions on android

### ISSUE THREAD

**Issue by longchr123** [tensorflow/tensorflow#20871]
**Question description：**
####  training a pb file, and then convert to tflite file, i use python testing pb file is correct ,but after into tflite file on android testing is wrong, I think this problem should be related to BN(tf.nn.batch_normalization),because when i remove BN get results of android and python are all the same, but with BN is different, the BN is tflite support,and the input data are the same, I don't know why?
------------------------------------

**Test Demo:**
```
def save_to_pb(sess):
    constant_graph = graph_util.convert_variables_to_constants(sess, sess.graph_def, ['out'])
    with tf.gfile.FastGFile(MODEL_DIR + 'expert-graph.pb', mode='wb') as f:
        f.write(constant_graph.SerializeToString())

def read_data(session):
    image_name = '182133.jpg'
    TEST_IMAGINE_PATH = '/home/leve/lcr/ClelebA/Celebra_crop_20w+_w128_dataset/images/test_100/'
    image = cv.imread(TEST_IMAGINE_PATH + image_name)
    image = tf.reshape(image, [128, 128, 3])
    image = image.eval(session=session)
    image = image[np.newaxis, :]
    return image

def batch_norm_lite(x, train=True, bn_decay=0.5,epsilon = 0.001,name='bn'):
    is_training = tf.convert_to_tensor(train,dtype='bool',name='is_training')
    x_shape = x.get_shape()
    params_shape = x_shape[-1:]
    axis = list(range(len(x_shape) - 1))
    beta = tf.get_variable(name+'_beta', params_shape, initializer=tf.zeros_initializer())
    gamma = tf.get_variable(name+'_gamma', params_shape, initializer=tf.ones_initializer())
    moving_mean = tf.get_variable(name+'_moving_mean', params_shape, initializer=tf.zeros_initializer(), trainable=False)
    moving_variance = tf.get_variable(name+'_moving_variance', params_shape, initializer=tf.ones_initializer(), trainable=False)
    mean, variance = tf.nn.moments(x, axis)
    update_moving_mean = moving_averages.assign_moving_average(moving_mean, mean, bn_decay)
    update_moving_variance = moving_averages.assign_moving_average(moving_variance, variance, bn_decay)
    tf.add_to_collection(name+'_update_moving_mean', update_moving_mean)
    tf.add_to_collection(name+'_update_moving_variance', update_moving_variance)
    mean, variance = control_flow_ops.cond(
        is_training, lambda: (mean, variance),
        lambda: (moving_mean, moving_variance))

    return tf.nn.batch_normalization(x, mean, variance, beta, gamma, epsilon,name=name)

def train_model():
    img = tf.placeholder(name="img", dtype=tf.float32, shape=(1, 128, 128, 3))
    b = tf.Variable(tf.truncated_normal((1, 128, 128, 3), seed=1),name='w1')
    y_real = img + tf.constant([1., 2., 3.]) + tf.constant([1., 4., 4.])
    val = tf.add(img, b)
    val = batch_norm_lite(val)
    out = tf.identity(val, name="out")
    MSE = tf.reduce_mean(tf.square(y_real - out), name='mse')
    train_step = tf.train.GradientDescentOptimizer(0.9).minimize(MSE)
    saver = tf.train.Saver(max_to_keep=10)
    os.environ["CUDA_VISIBLE_DEVICES"] = ''
    config = tf.ConfigProto()
    config.gpu_options.per_process_gpu_memory_fraction = 0.1
    config.gpu_options.allow_growth = True
    with tf.Session() as sess:
        sess.run(tf.initialize_all_variables())
        for epoch in range(100):
            sess.run(train_step,feed_dict={img:read_data(sess)})
            if epoch % 20 == 0:
                save_path = saver.save(sess, MODEL_DIR + MODEL_NAME, global_step=epoch + 1)
            print('step = ' + str(epoch))
        save_to_pb(sess)
train_model()
```
-------------------------

#### run upper  code and get pb file, when you test pb file with python you maybe get a result like this:
39.09046 45.927864 84.797905 45.952957 62.701195 68.00796 41.789146 80.99372 81.67459 81.2203....

#### then run this command to create tflite file:
```
bazel run --config=opt tensorflow/contrib/lite/toco:toco -- --input_file=/path/expert-graph.pb --output_file=/path/expert-graph.tflite --input_format=TENSORFLOW_GRAPHDEF --output_format=TFLITE --input_shape=1,128,128,3 --input_array=img --output_array=out --inference_type=FLOAT --input_data_type=FLOAT --allow_custom_ops
```

#### you will get expert-graph.tflite,then use tflite file to android,will get result like this:
49.392387， 42.290344，135.30707， 7.7764254 ， 125.98051，79.317825 ，157.29184 ，107.58522 ， 283.46997 ，36.103508....

#### Android and python have the same input data,but the result is different,if i remove BN and train, that's ok,but BN is very important,i can remove it ,i don't know how to solve it,this question has been around me for a long time,please help solve it, if don't use TF Lite, what should i do?Thanks
------------------------

### System information
Linux Ubuntu 16.04:
TensorFlow installed from source:
TensorFlow version 1.8.0:
Python version:3.6:
Bazel version 0.11.1:
GCC/Compiler version 5.4.0 20160609 (Ubuntu 5.4.0-6ubuntu1~16.04.10) :
CUDA/cuDNN 9.0/7.0.5:
GPU 1060-6G:

**Comment by shashishekhar** [comment]
@longchr123  Can you include the test data and your implementation of custom op: SquaredDifference ?

**Comment by longchr123** [comment]
@shashishekhar  there is no test data for training, If i don't use --allow_custom_ops in bazel command,there will be errors,like this:
```
Running command line: bazel-bin/tensorflow/contrib/lite/toco/toco '--input_file=/home/leve/lcr/model_result/tflite_test/expert-graph.pb' '--output_file=/home/leve/lcr/model_result/tflite_test/expert-graph.tflite' '--input_format=TENSORFLOW_GRAPHDEF' '--output_format=TFLITE' '--input_shape=1,128,128,3' '--input_array=img' '--output_array=out' '--inference_type=FLOAT' '--input_data_type=FLOAT'
2018-07-18 10:50:45.979461: I tensorflow/contrib/lite/toco/import_tensorflow.cc:1265] Converting unsupported operation: SquaredDifference
2018-07-18 10:50:45.979688: I tensorflow/contrib/lite/toco/graph_transformations/graph_transformations.cc:39] Before Removing unused ops: 27 operators, 41 arrays (0 quantized)
2018-07-18 10:50:45.979826: I tensorflow/contrib/lite/toco/graph_transformations/graph_transformations.cc:39] Before general graph transformations: 27 operators, 41 arrays (0 quantized)
2018-07-18 10:50:45.980017: I tensorflow/contrib/lite/toco/graph_transformations/graph_transformations.cc:39] After general graph transformations pass 1: 13 operators, 21 arrays (0 quantized)
2018-07-18 10:50:45.980086: I tensorflow/contrib/lite/toco/graph_transformations/graph_transformations.cc:39] Before dequantization graph transformations: 13 operators, 21 arrays (0 quantized)
2018-07-18 10:50:45.980144: I tensorflow/contrib/lite/toco/allocate_transient_arrays.cc:329] Total transient array allocated size: 196736 bytes, theoretical optimal value: 196672 bytes.
2018-07-18 10:50:45.980271: F tensorflow/contrib/lite/toco/tflite/export.cc:315] Some of the operators in the model are not supported by the standard TensorFlow Lite runtime. If you have a custom implementation for them you can disable this error with --allow_custom_ops. Here is a list of operators for which you will need custom implementations: RSQRT, SquaredDifference.
```

**how to sovle it?**

**Comment by shashishekhar** [comment]
Ah, I am going to explain what the error means:
First some background, TFLite doesn't have all the operators available in Tensorflow, we are working on improving and adding more operator support. When you convert your model the tool tells you if there are operators which don't have implementations. Here the tool is mentioning that RSQRT, SquaredDifference may need custom implementations. 

I am surprised the model didn't crash, I will take a look and see if these ops were used only for training, usually toco strips out the ops used for training.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @shashishekhar: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.
